titus-supplementary-component/tasks-publisher/src/main/java/com/netflix/titus/supplementary/taskspublisher/TasksPublisherCtrl.java

```diff
@@ -0,0 +1,183 @@
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.api.patterns.PolledMeter;
+import com.netflix.titus.api.jobmanager.JobAttributes;
+import com.netflix.titus.common.util.rx.RetryHandlerBuilder;
+import com.netflix.titus.common.util.tuple.Pair;
+import com.netflix.titus.ext.elasticsearch.TaskDocument;
+import com.netflix.titus.grpc.protogen.Job;
+import com.netflix.titus.grpc.protogen.Task;
+import com.netflix.titus.runtime.endpoint.v3.grpc.V3GrpcModelConverters;
+import org.eclipse.jetty.util.ConcurrentHashSet;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+
+public class TasksPublisherCtrl {
+    private static final Logger logger = LoggerFactory.getLogger(TasksPublisherCtrl.class);
+    public static final String TASK_DOCUMENT_CONTEXT = "taskDocumentContext";
+    private final EsClient esClient;
+    private Map<String, String> taskDocumentBaseContext;
+    private Registry registry;
+    private TitusClient titusClient;
+
+    private AtomicInteger numErrors = new AtomicInteger(0);
+    private AtomicInteger numIndexUpdated = new AtomicInteger(0);
+    private AtomicInteger numTasksUpdated = new AtomicInteger(0);
+    private Set<String> uniqueTasksIdUpdated = new ConcurrentHashSet<>();
+    private AtomicLong lastPublishedTimestamp;
+
+    private static final long INITIAL_RETRY_DELAY_MS = 500;
+    private static final long MAX_RETRY_DELAY_MS = 2_000;
+    private Disposable subscription;
+
+    public TasksPublisherCtrl(EsClient esClient,
+                              TitusClient titusClient,
+                              @Qualifier(TASK_DOCUMENT_CONTEXT) Map<String, String> taskDocumentBaseContext,
+                              Registry registry) {
+        this.esClient = esClient;
+        this.titusClient = titusClient;
+        this.taskDocumentBaseContext = taskDocumentBaseContext;
+        this.registry = registry;
+        configureMetrics();
+    }
+
+
+    public AtomicInteger getNumErrors() {
+        return numErrors;
+    }
+
+    public AtomicInteger getNumIndexUpdated() {
+        return numIndexUpdated;
+    }
+
+    public AtomicInteger getNumTasksUpdated() {
+        return numTasksUpdated;
+    }
+
+    @PostConstruct
+    public void start() {
+        subscription = titusClient.getTaskUpdates()
+                .publishOn(Schedulers.elastic())
+                .map(task -> {
+                    final Mono<Job> jobById = titusClient.getJobById(task.getJobId());
+                    return Pair.of(task, jobById);
+                })
+                .flatMap(taskMonoPair -> {
+                    final Task task = taskMonoPair.getLeft();
+                    return taskMonoPair.getRight()
+                            .map(job -> {
```

Comment:
Fetching job for each task update is very expensive, especially that job data are pushed via stream as well. Direct call should be only a fallback strategy.